packages/loader/container-loader/src/retriableDocumentStorageService.ts

```diff
@@ -39,22 +51,35 @@ export class RetriableDocumentStorageService implements IDocumentStorageService,
 	}
 
 	public get repositoryUrl(): string {
-		return this.internalStorageService.repositoryUrl;
+		if (this.internalStorageService) {
+			return this.internalStorageService.repositoryUrl;
+		}
+		throw new Error("storage service not yet instantiated");
 	}
 
 	public async getSnapshotTree(
 		version?: IVersion,
 		scenarioName?: string,
 	): Promise<ISnapshotTree | null> {
+		if (this.internalStorageService === undefined) {
+			await this.storageServiceDeferred.promise;
+		}
 		return this.runWithRetry(
-			async () => this.internalStorageService.getSnapshotTree(version, scenarioName),
+			async () =>
```

Comment:
```suggestion
			async () =>
				this.internalStorageServiceP.then(async (s)=>s.getSnapshotTree(
					version,
					scenarioName,
				)),```

Reply:
do this for all the methods and remove the deferred

Reply:
also, casting is bad, it should almost never be done, and when done it should be limit to as a few places as possible